packages/experimental-utils/src/ts-eslint/SourceCode.ts

```diff
@@ -175,6 +172,10 @@ namespace SourceCode {
     tokenOrComment: TSESTree.Token | TSESTree.Comment,
   ) => boolean;
 
+  export type ReturnTypeFromOptions<T> = T extends { includeComments: true }
+    ? TSESTree.Token | TSESTree.Comment
+    : TSESTree.Token;
```

Comment:
this type is better than previous one but its still not taking into account filtering from filter property if its type guard 

for filter from object options and when passed as function we should use simmilar logic as in `array.filter` from `lib`

Reply:
would this be better as this, with a generic constraint?

```ts
export type ReturnTypeFromOptions<
  T extends SourceCode.CursorWithCountOptions
> = T['includeComments'] extends true
  ? TSESTree.Token | TSESTree.Comment
  : TSESTree.Token;
```

Reply:
CursorWithCountOptions is object, number or function and we have to validate first if T is object, this is not going to work
> Type '"includeComments"' cannot be used to index type 'T'.

Reply:
additionally now T can be `CursorWithCountOptions` or `CursorWithSkipOptions`